### ISSUE TITLE
Implement pooling backward tester in layer_testers.mojo

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#3720]
The audit for #3208 confirmed that no pooling backward tester exists in `shared/testing/layer_testers.mojo`. When MaxPool/AvgPool backward passes are implemented, a `test_pool_layer_backward` (or equivalent) should be added following the same pattern as conv2d/batchnorm backward testers, with appropriate epsilon/tolerance documented per dtype.

_Follow-up from #3208_

**Comment by mvillmow** [comment]
# Implementation Plan

Now I have all the information needed. Here is the implementation plan:

---

## Implementation Plan: Issue #3720 — Pooling Backward Tester in `layer_testers.mojo`

### 1. Objective

Add `test_pool_layer_backward` (and `test_maxpool_layer_backward` / `test_avgpool_layer_backward` as needed) to `shared/testing/layer_testers.mojo`, following the same structure as `test_conv_layer_backward` and `test_batchnorm_layer_backward`. The tester must validate:
- Gradient shape correctness
- Gradient routing (MaxPool: only max positions receive gradient)
- Gradient distribution (AvgPool: equal redistribution)
- Numerical gradient matching via `check_gradient` / `compute_numerical_gradient`

---

### 2. Approach

**Key decisions:**

- Add a **single `test_pool_layer_backward` static method** with a `pool_type: String` parameter (`"max"` or `"avg"`), mirroring `test_pooling_layer`'s API. Avoids duplicating the tester scaffolding.
- Follow the **three-tier structure**: shape assertion → analytical value check → numerical gradient check.
- Use **non-uniform `grad_output`** (e.g., `Float32(i % 4) * 0.25 - 0.3`) to avoid AvgPool cancellation (per `batch-norm-gradient-test-fix` lesson).
- MaxPool: include a **gradient routing test** (verify 0-gradient at non-max positions) for the analytic check tier.
- AvgPool: include a **gradient distribution test** (verify equal distribution within window) for the analytic check tier.
- Use **epsilon = `GRADIENT_CHECK_EPSILON_FLOAT32` (3e-4)** for float32, `GRADIENT_CHECK_EPSILON_OTHER` for other dtypes.
- Use **`rtol=1e-3, atol=5e-4`** for pool layers (matches `test_backward_conv_pool.mojo` existing values).
- Keep test input **small** (`(2, 2, 4, 4)` with 2×2 pool, stride 2) for O(n²) gradient checking tractability.
- The pooling backward functions return bare `ExTensor`, not a gradient container struct — access directly (no `.grad_input` indirection).

---

### 3. Files to Create

**None.** The new function is added to an existing file.

---

### 4. Files to Modify

#### `shared/testing/layer_testers.mojo` (primary)

Add a new `@staticmethod` method after `test_batchnorm_layer_backward` (line 1156):

```
fn test_pool_layer_backward(
    channels: Int,
    input_h: Int,
    input_w: Int,
    pool_size: Int,
    stride: Int,
    dtype: DType,
    pool_type: String = "max",
    padding: Int = 0,
) raises
```

Structure of the new method:

1. **Shape tier** — Create `(2, channels, input_h, input_w)` seeded random input; run forward + backward; assert `grad_input.shape() == input.shape()`.
2. **Analytical value tier** (pool-type-specific):
   - *MaxPool*: Use 1×1 batch, 1 channel, 2×2 input with known values `[1, 4, 2, 3]`; kernel=2, stride=2. Run forward + backward with uniform `grad_output=1.0`. Assert gradient is 1.0 only at position of `4.0`, 0.0 elsewhere.
   - *AvgPool*: Use 1×1 batch, 1 channel, 2×2 input with ones; kernel=2, stride=2. Run forward + backward with `grad_output=1.0`. Assert each input position receives `0.25`.
3. **Numerical gradient tier** — Use `(1, 2, 4, 4)` seeded input; non-uniform `grad_output` pattern; call `compute_numerical_gradient` with `forward_for_grad = fn(x) -> sum(output * grad_output)`. Compare against `maxpool2d_backward` / `avgpool2d_backward` result using `assert_gradients_close(rtol=1e-3, atol=5e-4)`.

**Important implementation notes:**
- `forward_for_grad` must return a scalar (sum of weighted output) so `compute_numerical_gradient` computes `dL/dx` correctly.
- `grad_output` tensor shape = output shape from pool forward pass; use `pool_output_shape()` helper already imported in the file.
- Non-uniform pattern for `grad_output`: iterate `output.numel()`, set element `i` to `Float32(i % 4) * 0.25 - 0.3` to avoid cancellation.

**Docstring must document:**
- epsilon = `3e-4` for float32 (consistent with issue #2704)
- tolerance = `rtol=1e-3, atol=5e-4` for pool layers
- Non-uniform `grad_output` rationale (cancellation avoidance for AvgPool)

---

### 5. Implementation Order

1. **Read** `shared/testing/layer_testers.mojo` lines 1–100 (imports section) to confirm `maxpool2d_backward`, `avgpool2d_backward`, and `pool_output_shape` are already imported. If not, add imports.
2. **Read** full `test_pooling_layer` method (lines 255–341) and `test_batchnorm_layer_backward` (lines 1078–1156) one more time for exact formatting and code style reference.
3. **Write** `test_pool_layer_backward` function body in three tiers (shape → analytic → numerical), appended after line 1156.
4. **Run pre-commit hooks** with `SKIP=mojo-format` if on a host with GLIBC < 2.32 (CI will run `mojo format`).
5. **Verify** by running the existing tests in `tests/shared/core/test_backward_conv_pool.mojo` to ensure no regressions:
   ```bash
   just test-group "tests/shared/core" "test_backward_conv_pool.mojo"
   ```
6. **Create PR** on branch `3720-pool-backward-tester` with `Closes #3720`.

---

### 6. Verification

**Correctness checks:**
- `mojo build shared/testing/layer_testers.mojo` compiles without errors.
- All existing tests in `test_backward_conv_pool.mojo` still pass (no regression).
- Manually call `LayerTester.test_pool_layer_backward(2, 4, 4, 2, 2, DType.float32, "max")` and `"avg"` in a scratch test file to confirm output.

**Expected behavior:**
- MaxPool routing test: gradient is nonzero exactly at the max-position index, zero elsewhere.
- AvgPool distribution test: each input cell within window receives `1.0 / (pool_size * pool_size)`.
- Numerical gradient check: relative error < `1e-3` for all sampled positions with float32.

**CI validation:**
- Push branch and confirm `.github/workflows/comprehensive-tests.yml` passes for the `shared/core` test group.

---

### 7. Skills Used

**Invoked during planning:**
- `docs-search` — searched for existing backward tester patterns and gradient checking utilities in the codebase
- `graph-query` — traced function dependencies between `layer_testers.mojo`, `pooling.mojo`, and `gradient_checker.mojo`

**Team knowledge base skills referenced (from Prior Learnings):**

| Skill | How It Shaped This Plan |
|-------|------------------------|
| `batch-norm-gradient-test-fix` | Adopted non-uniform `grad_output` pattern to prevent AvgPool cancellation (same symmetry cancellation risk as BatchNorm) |
| `reenable-disabled-mojo-backward-tests` | Used GradientTriple field naming verification approach; confirmed pooling backward returns bare `ExTensor` (no `.grad_input` wrapper) |
| `batch-norm-backward-gradient-analysis` | Confirmed three-tier test structure (shape → analytic → numerical) as the canonical pattern for new backward testers |

---
*Generated by Claude Code Planner*
